### DISCUSSION TITLE
How to draw the object outlines using Silk.NET?

### DISCUSSION THREAD

**Discussion by Hanning-Liu** [dotnet/Silk.NET#1834] [Q&A] [answered]
I want to implement the effect like the right side of the screenshot below:
Is there any method to implement this outline drawing effect using Silk.NET?
![744d5e6732aaf8fe8795b482db99634](https://github.com/dotnet/Silk.NET/assets/42676581/346d6566-e969-4934-b200-8c3debb57d21)

**Comment by Beyley** [comment]
This would generally be done by writing a custom shader, this functionality is not provided by Silk.NET itself, since "outlines" arent really a concept of graphics APIs
Heres a repository showing off some ways of doing outlines with OpenGL
https://github.com/zwcloud/OpenGLOutline

**Reply by Hanning-Liu** [reply]
Yes, I've seen the code in this repository yesterday! And I find it cannot correctly handle the scenario with multiple geometries at front and behind.

**Reply by Beyley** [reply]
Have you sorted your geometry by depth?

**Reply by qian-o** [reply]
https://learnopengl.com/Advanced-OpenGL/Stencil-testing

Perform template tests on each mesh group to outline the general contour.

**Reply by Hanning-Liu** [reply]
Sorry, I don't know how to do the sort operation. And CPP code is hard for me to read. 😔

**Reply by Hanning-Liu** [reply]
Hi @Beyley  and @qian-o ,
Thank you for your help! I will learn the post from learnopengl.com carefully, and hope I can handle the cpp codes!

**Reply by qian-o** [reply, marked as answer]
I have written similar stencil testing code before and I hope it can help you.

![image](https://github.com/dotnet/Silk.NET/assets/84434846/fc462836-b8f3-4fe4-b19b-d16a1d92d511)

https://github.com/qian-o/LearnOpenGL/blob/master/4_2_StencilTesting/Program.cs

**Reply by Hanning-Liu** [reply]
Wow, thank you! It will be very helpful to me!!!

**Reply by Hanning-Liu** [reply]
Hi @qian-o ,
I find the method on the website is to scale every object. And it will lead to the different outline thickness between every object.
![image](https://github.com/dotnet/Silk.NET/assets/42676581/f074a8a6-31d9-4b27-ba73-223d74d0bb6a)
![image](https://github.com/dotnet/Silk.NET/assets/42676581/63f5a3a5-fbc2-4854-b9a1-b54595d95e84)

**Reply by qian-o** [reply]
Yes, the outline is achieved through scaling. 
If you have any further questions, you can join the official [Discord](https://discord.com/invite/DTHHXRt) for discussion or contact me privately through the contact information on my profile page.
